master/buildbot/status/status_stash.py

```diff
@@ -0,0 +1,141 @@
+        self.body = body
+        self.length = len(body)
+
+    def startProducing(self, consumer):
+        consumer.write(self.body)
+        return defer.succeed(None)
+
+    def pauseProducing(self):
+        pass
+
+    def stopProducing(self):
+        pass
+
+
+class StashStatusPush(StatusReceiverMultiService,
+                      BuildSetSummaryNotifierMixin):
+    implements(IStatusReceiver)
+
+    def __init__(self, host, user, password):
+        """
+        :param host: The domain name or IP of the stash host.
+        :param user: The stash user to log in as using http basic auth.
+        :param password: The password to use with the stash user.
+        :return:
+        """
+        StatusReceiverMultiService.__init__(self)
+        self.base_url = 'https://%s/rest/build-status/1.0/commits/' % (host, )
+        self.auth = b64encode('%s:%s' % (user, password))
+        self._sha = Interpolate('%(src::revision)s')
+        self.master_status = None
+
+    @defer.inlineCallbacks
+    def send(self, builderName, build, status):
+        (sha, ) = yield defer.gatherResults([build.render(self._sha), ])
+        build_url = build.builder.status.getURLForThing(build)
+        body = dumps({'state': status, 'key': builderName, 'url': build_url})
+        stash_uri = self.base_url + sha
+        agent = Agent(reactor)
+        d = agent.request(
+            method='POST',
+            uri=bytes(stash_uri),
+            headers=Headers({
+                'Content-Type': ['application/json; charset=utf-8',],
+                'Authorization': ['Basic %s' % (self.auth, ), ],
+                'Accept': ['*/*', ],
+                'Accept-Language': ['en-US, en;q=0.8', ],
+            }),
+            bodyProducer=StringProducer(body)
+        )
+        d.addCallback(logIfNot2XX)
+        d.addErrback(log.err,
+                     'StashStatusPush failed while POSTing status message '
+                     'for commit %s built by builder %s to %s with body %s'
+                     '' % (sha, builderName, stash_uri, body))
+
+    def setServiceParent(self, parent):
+        """
+        @type  parent: L{buildbot.master.BuildMaster}
+        """
+        StatusReceiverMultiService.setServiceParent(self, parent)
+        self.master_status = self.parent
+        self.master_status.subscribe(self)
+        self.master = self.master_status.master
+
+    def startService(self):
+        print """Starting up StashStatusPush"""
+        self.summarySubscribe()
+        StatusReceiverMultiService.startService(self)
+
+    def stopService(self):
+        self.summaryUnsubscribe()
+
+    def builderAdded(self, name, builder):
+        return self
+
+    def buildStarted(self, builderName, build):
+        self.send(builderName, build, INPROGRESS)
+        return self
+
+    def buildFinished(self, builderName, build, results):
+        self.send(builderName, build,
+                  SUCCESSFUL if results == SUCCESS else FAILED)
```

Comment:
@sa2ajj this doesn't appear to be working. I have builds that are passing that are being reported as FAILED. So... another issue. I'll look into it now.